### PR TITLE
[FLINK-12801][table-planner-blink] set parallelism for batch SQL

### PR DESCRIPTION
**Pull request by XuPingyong**
What is the purpose of the change

DataStream user can set parallelism by SingleOutputStreamOperator#setParallelism and DataStreamSink#setParallelism. But SQL users cannot set parallelism  to operators while compiled jobGraphs from SQL are usally complex.
       Now we first set parallelism for batch SQL by config. We introduce two resourceSetting mode:
       InferMode.NONE:  User can set parallelism to source, sink and other nodes separately.
       InferMode.ONLY_SOURCE： Relative to  InferMode.NONE, source paralelism can be inferred by source row count.
        We also introduce ShuffleStage to make adjacent operatos parallelism same that there is no data shuffle between them.

Brief change log

Introduce parallelism setting for batch SQL

Verifying this change

ut

Does this pull request potentially affect one of the following parts:

Dependencies (does it add or upgrade a dependency): (no)
The public API, i.e., is any changed class annotated with @Public(Evolving): (no)
The serializers: (no)
The runtime per-record code paths (performance sensitive): (no)
Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Yarn/Mesos, ZooKeeper: (no)
The S3 file system connector: (no)
Documentation

Does this pull request introduce a new feature? (yes)
If yes, how is the feature documented? (JavaDocs)

### CODE REVIEW

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/plan/nodes/resource/batch/parallelism/BatchParallelismProcessor.java [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+	private TableEnvironment tEnv;
+
+	@Override
+	public List<ExecNode<?, ?>> process(List<ExecNode<?, ?>> sinkNodes, DAGProcessContext context) {
+		sinkNodes.forEach(s -> Preconditions.checkArgument(s instanceof BatchExecNode));
+		tEnv = context.getTableEnvironment();
+		List<ExecNode<?, ?>> rootNodes = filterSinkNodes(sinkNodes);
+		Map<ExecNode<?, ?>, Integer> nodeToFinalParallelismMap = BatchFinalParallelismSetter.calculate(tEnv, rootNodes);
+		Map<ExecNode<?, ?>, ShuffleStage> nodeShuffleStageMap = ShuffleStageGenerator.generate(rootNodes, nodeToFinalParallelismMap);
+		new BatchShuffleStageParallelismCalculator(tEnv.getConfig().getConf(), tEnv.streamEnv().getParallelism());
```
Comment (JingsongLi) [inline flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/plan/nodes/resource/batch/parallelism/BatchParallelismProcessor.java]: Forgot to call calculate?

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/plan/nodes/resource/batch/parallelism/BatchParallelismProcessor.java [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Processor for calculating parallelism for {@link BatchExecNode}.
+ */
+public class BatchParallelismProcessor implements DAGProcessor {
+
+	private TableEnvironment tEnv;
```
Comment (JingsongLi) [inline flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/plan/nodes/resource/batch/parallelism/BatchParallelismProcessor.java]: Be a local variable?

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/api/BatchTableEnvironment.scala [hunk 3]
```diff
@@ -121,7 +123,9 @@ class BatchTableEnvironment(
   override private[flink] def translateToExecNodeDag(rels: Seq[RelNode]): Seq[ExecNode[_, _]] = {
     val nodeDag = super.translateToExecNodeDag(rels)
     // breakup deadlock
-    new DeadlockBreakupProcessor().process(nodeDag)
+    val postNodeDag = new DeadlockBreakupProcessor().process(nodeDag)
```
Comment (JingsongLi) [inline flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/api/BatchTableEnvironment.scala]: Looks `DeadlockBreakupProcessor` is a `DAGProcessor` too. Just let it implement `DAGProcessor` too?

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/plan/nodes/resource/batch/parallelism/BatchParallelismProcessor.java [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+		Map<ExecNode<?, ?>, Integer> nodeToFinalParallelismMap = BatchFinalParallelismSetter.calculate(tEnv, rootNodes);
+		Map<ExecNode<?, ?>, ShuffleStage> nodeShuffleStageMap = ShuffleStageGenerator.generate(rootNodes, nodeToFinalParallelismMap);
+		new BatchShuffleStageParallelismCalculator(tEnv.getConfig().getConf(), tEnv.streamEnv().getParallelism());
+		for (ExecNode<?, ?> node : nodeShuffleStageMap.keySet()) {
+			node.getResource().setParallelism(nodeShuffleStageMap.get(node).getParallelism());
+		}
+		return sinkNodes;
+	}
+
+	private List<ExecNode<?, ?>> filterSinkNodes(List<ExecNode<?, ?>> sinkNodes) {
```
Comment (JingsongLi) [inline flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/plan/nodes/resource/batch/parallelism/BatchParallelismProcessor.java]: Why filter sink node? Add reason comments to `BatchParallelismProcessor`?

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/plan/nodes/resource/batch/parallelism/ShuffleStageGenerator.java [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+			// source node
+			ShuffleStage shuffleStage = new ShuffleStage();
+			shuffleStage.addNode(execNode);
+			if (nodeToFinalParallelismMap.containsKey(execNode)) {
+				shuffleStage.setParallelism(nodeToFinalParallelismMap.get(execNode), true);
+			}
+			nodeShuffleStageMap.put(execNode, shuffleStage);
+		} else if (execNode instanceof BatchExecExchange && !(((BatchExecExchange) execNode).getDistribution().getType() == RelDistribution.Type.RANGE_DISTRIBUTED)) {
+			// do nothing
+		} else if (execNode instanceof BatchExecSink) {
```
Comment (JingsongLi) [inline flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/plan/nodes/resource/batch/parallelism/ShuffleStageGenerator.java]: Here you consider `BatchExecSink`, and you exclude `BatchExecSink` before?

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/plan/nodes/resource/batch/parallelism/ShuffleStageGenerator.java [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+
+		if (execNode.getInputNodes().isEmpty()) {
+			// source node
+			ShuffleStage shuffleStage = new ShuffleStage();
+			shuffleStage.addNode(execNode);
+			if (nodeToFinalParallelismMap.containsKey(execNode)) {
+				shuffleStage.setParallelism(nodeToFinalParallelismMap.get(execNode), true);
+			}
+			nodeShuffleStageMap.put(execNode, shuffleStage);
+		} else if (execNode instanceof BatchExecExchange && !(((BatchExecExchange) execNode).getDistribution().getType() == RelDistribution.Type.RANGE_DISTRIBUTED)) {
```
Comment (JingsongLi) [inline flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/plan/nodes/resource/batch/parallelism/ShuffleStageGenerator.java]: line too long, please enter a new line.

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/plan/nodes/resource/batch/parallelism/ShuffleStageGenerator.java [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static java.util.stream.Collectors.toList;
+
+/**
+ * Build shuffleStages.
+ */
+public class ShuffleStageGenerator {
```
Comment (JingsongLi) [inline flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/plan/nodes/resource/batch/parallelism/ShuffleStageGenerator.java]: Can you add some comments to explain how to shuffle stage divide?

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/plan/nodes/physical/batch/BatchExecSortWindowAggregate.scala [hunk 8]
```diff
@@ -96,8 +96,4 @@ class BatchExecSortWindowAggregate(
     val aggregateNamePrefix = if (isMerge) "Global" else "Complete"
     aggregateNamePrefix + "WindowSortAggregateBatchExec"
   }
-
-  override def getParallelism(input: StreamTransformation[BaseRow], conf: TableConfig): Int = {
-    if (isFinal && grouping.length == 0) 1 else input.getParallelism
```
Comment (JingsongLi) [inline flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/plan/nodes/physical/batch/BatchExecSortWindowAggregate.scala]: How do you set parallelism 1 in this case? Maybe you can add some comments in `BatchShuffleStageParallelismCalculator`?

File: flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/plan/nodes/resource/batch/parallelism/ShuffleStageGenerator.java (hunk 5 above)
Comment (XuPingyong) [reply flink-table/flink-table-planner-blink/src/main/java/org/apache/flink/table/plan/nodes/resource/batch/parallelism/ShuffleStageGenerator.java]: > Here you consider `BatchExecSink`, and you exclude `BatchExecSink` before?

Double exclude, I fix it.

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/plan/nodes/physical/batch/BatchExecSortWindowAggregate.scala (hunk 8 above)
Comment (XuPingyong) [reply flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/plan/nodes/physical/batch/BatchExecSortWindowAggregate.scala]: > How do you set parallelism 1 in this case? Maybe you can add some comments in `BatchShuffleStageParallelismCalculator`?

I add comments in BatchFinalParallelismSetter, as its parallelism is set to 1 depends on the global exchange.

File: flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/plan/metadata/FlinkRelMdRowCount.scala [hunk 9]
```diff
@@ -171,7 +171,8 @@ class FlinkRelMdRowCount private extends MetadataHandler[BuiltInMetadata.RowCoun
     } else {
       val inputRowCnt = mq.getRowCount(input)
       val config = rel.getCluster.getPlanner.getContext.asInstanceOf[FlinkContext].getTableConfig
-      val parallelism = NodeResourceConfig.calOperatorParallelism(inputRowCnt, config.getConf)
+      val parallelism = (inputRowCnt /
+          config.getConf.getLong(PlannerConfigOptions.SQL_OPTIMIZER_ROWS_PER_LOCALAGG) + 1).toInt
```
Comment (JingsongLi) [inline flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/plan/metadata/FlinkRelMdRowCount.scala]: Explain this change?

Comment (XuPingyong) [reply flink-table/flink-table-planner-blink/src/main/scala/org/apache/flink/table/plan/metadata/FlinkRelMdRowCount.scala]: LocalAgg degree was depend on parallelism calculating that happens after plan optimizing, it is unreasonable. So it predicts parallelism itself, that cause no conflict.

**Comment by JingsongLi** [comment]
LGTM +1

**Pull request closed without merging**